cmd/encryption.go

```diff
@@ -0,0 +1,115 @@
+package cmd
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/base64"
+	"errors"
+	"io"
+	"net/http"
+
+	"github.com/minio/sio"
+)
+
+var (
+	errInsecureSSERequest  = errors.New("Requests specifying Server Side Encryption with Customer provided keys must be made over a secure connection")
+	errInvalidSSEAlgorithm = errors.New("Requests specifying Server Side Encryption with Customer provided keys must provide a valid encryption algorithm")
+	errMissingSSEKey       = errors.New("Requests specifying Server Side Encryption with Customer provided keys must provide an appropriate secret key")
+	errInvalidSSEKey       = errors.New("The secret key was invalid for the specified algorithm")
+	errMissingSSEKeyMD5    = errors.New("Requests specifying Server Side Encryption with Customer provided keys must provide the client calculated MD5 of the secret key")
+	errSSEKeyMD5Mismatch   = errors.New("The calculated MD5 hash of the key did not match the hash that was provided")
+)
+
+func isSSERequest(header http.Header) bool {
+	if header.Get("X-Amz-Server-Side-Encryption-Customer-Algorithm") != "" {
+		return true
+	}
+	if header.Get("X-Amz-Server-Side-Encryption-Customer-Key") != "" {
+		return true
+	}
+	if header.Get("X-Amz-Server-Side-Encryption-Customer-Key-MD5") != "" {
+		return true
+	}
+	return false
+}
+
+func parseSSERequest(header http.Header) ([]byte, error) {
+	algorithm := header.Get("X-Amz-Server-Side-Encryption-Customer-Algorithm")
+	if algorithm != "AES256" {
+		return nil, errInvalidSSEAlgorithm
+	}
+	key, err := base64.StdEncoding.DecodeString(header.Get("X-Amz-Server-Side-Encryption-Customer-Key"))
+	if err != nil {
+		return nil, errInvalidSSEKey
+	}
+	keyMD5, err := base64.StdEncoding.DecodeString(header.Get("X-Amz-Server-Side-Encryption-Customer-Key-MD5"))
+	if err != nil {
+		return nil, errSSEKeyMD5Mismatch
+	}
+	if md5Sum := md5.Sum(key); !bytes.Equal(md5Sum[:], keyMD5) {
+		return nil, errSSEKeyMD5Mismatch
+	}
+	return key, nil
+}
+
+type sseObject struct {
+	ObjectLayer
+
+	key []byte
+}
+
+func (sse sseObject) GetObjectInfo(bucket string, object string) (ObjectInfo, error) {
+	info, err := sse.ObjectLayer.GetObjectInfo(bucket, object)
+	if err != nil {
+		return ObjectInfo{}, err
+	}
+	size := (info.Size / (32 + 64*1024)) * (64 * 1024)
+	if mod := info.Size % (32 + 64*1024); mod > 0 {
+		size += mod - 32
```

Comment:
Can mod ever be smaller than 32? Then you would effectively decrease the size...